language/move-core/types/src/transaction_argument.rs

```diff
@@ -13,6 +13,7 @@ pub enum TransactionArgument {
     Address(AccountAddress),
     U8Vector(#[serde(with = "serde_bytes")] Vec<u8>),
     Bool(bool),
+    Vector(Vec<TransactionArgument>),
```

Comment:
it seems like this would allow heterogenous vectors, like `[U8(4), U64(123), Address(..), ..]`, which seems wrong i guess.

also, do we want to disallow the generic `Vector(U8)`? it definitely seems a bit weird to have multiple encodings of the same thing.

do we need any hardening against deeply nested vectors?

Reply:
yeah this would allow a `Struct` effectively but we think that may be inevitable and error'ed on conversation

Reply:
> do we need any hardening against deeply nested vectors?
we should be protected at few level, in las and in the VM, but certainly something to double check when we have the final solution in place

Reply:
> It seems like this would allow heterogenous vectors, like `[U8(4), U64(123), Address(..), ..]`, which seems wrong i guess.

This is equivalent to sending garbage LCS bytes. (Calling a Move script is a RPC protocol using LCS/TransactionArgument as transport.)

> also, do we want to disallow the generic `Vector(U8)`? it definitely seems a bit weird to have multiple encodings of the same thing.

LCS itself has two types actually (as well as the move vm). This is a bit annoying but it is worth it : many languages have a special type for byte-strings which is not the same as Vec<u8>. The latter is typically less efficient (and this is visible in this PR as well).

Anyways, it is up to Move converters to disallow it.

> do we need any hardening against deeply nested vectors?

LCS already has a limit in place (although I haven't finished updated the client-side runtimes).

Reply:
I carried on with the idea and explicitly do not attempt to type-check arguments during conversion to (tentative) move values.